src/server/http_routes.rs

```diff
@@ -133,33 +136,93 @@ pub enum ApplicationError {
-            Self::RouteNotFound { .. } => StatusCode::NOT_FOUND,
-            Self::DatabaseError { .. } => StatusCode::INTERNAL_SERVER_ERROR,
-            Self::JsonGenerationError { .. } => StatusCode::INTERNAL_SERVER_ERROR,
-        }
+    pub fn response(&self) -> Result<Response<Body>, Self> {
+        Ok(match self {
+            Self::BucketByName { .. } => self.internal_error(),
+            Self::BucketMappingError { .. } => self.internal_error(),
+            Self::WritingPoints { .. } => self.internal_error(),
+            Self::Query { .. } => self.internal_error(),
+            Self::QueryError { .. } => self.bad_request(),
+            Self::BucketNotFound { .. } => self.not_found(),
+            Self::RequestSizeExceeded { .. } => self.bad_request(),
+            Self::ExpectedQueryString { .. } => self.bad_request(),
+            Self::InvalidQueryString { .. } => self.bad_request(),
+            Self::InvalidRequestBody { .. } => self.bad_request(),
+            Self::InvalidContentEncoding { .. } => self.bad_request(),
+            Self::ReadingHeaderAsUtf8 { .. } => self.bad_request(),
+            Self::ReadingBody { .. } => self.bad_request(),
+            Self::ReadingBodyAsUtf8 { .. } => self.bad_request(),
+            Self::ParsingLineProtocol { .. } => self.bad_request(),
+            Self::ReadingBodyAsGzip { .. } => self.bad_request(),
+            Self::RouteNotFound { .. } => self.not_found(),
+            Self::DatabaseError { .. } => self.internal_error(),
+            Self::JsonGenerationError { .. } => self.internal_error(),
+        })
+    }
+
+    fn bad_request(&self) -> Response<Body> {
+        Response::builder()
+            .status(StatusCode::BAD_REQUEST)
+            .body(self.body())
+            .unwrap()
+    }
+
+    fn internal_error(&self) -> Response<Body> {
+        Response::builder()
+            .status(StatusCode::INTERNAL_SERVER_ERROR)
+            .body(self.body())
+            .unwrap()
+    }
+
+    fn not_found(&self) -> Response<Body> {
+        Response::builder()
+            .status(StatusCode::NOT_FOUND)
+            .body(Body::empty())
+            .unwrap()
+    }
+
+    fn body(&self) -> Body {
+        let json = serde_json::json!({"error": self.to_string()}).to_string();
+        Body::from(json)
     }
 }
 
 const MAX_SIZE: usize = 10_485_760; // max write request size of 10MB
 
+fn router<M: ConnectionManager + Send + Sync + Debug + 'static>(
+    server: Arc<AppServer<M>>,
+) -> Router<Body, ApplicationError> {
+    // Create a router and specify the the handlers.
+    Router::builder()
+        .data(server)
+        // this endpoint is for API backward compatibility with InfluxDB 2.x
+        .post("/api/v2/write", write::<M>)
+        .get("/ping", ping)
+        .get("/api/v2/read", read::<M>)
+        .get("/api/v1/partitions", list_partitions::<M>)
+        .post("/api/v1/snapshot", snapshot_partition::<M>)
```

Comment:
I think logging all requests (at least in debug level) as well as all errors is critical. I think we can do so via some middleware (e.g. the following). We can probably do something fancier with logging levels so that we always `info!` processing requests or something, but then `debug!` note when we are done with responses that are not errors, and error! ones that are errors... 

I recommend filing  a ticket to handle this in a subsequent PR

```suggestion
        .get("/api/v1/partitions", list_partitions::<M>)
        .post("/api/v1/snapshot", snapshot_partition::<M>)
                .middleware(Middleware::pre(|req| async move {
            info!(request = ?req, "Processing request");
        }))
        .middleware(Middleware::post(|res| async move {
            info!(response = ?res, "Successfully processed request");
            Ok(res)
        }))

```

Reply:
added in b998a916b3a7d566b0b02181bedbce5b03f5b444